planner/core/optimizer.go

```diff
@@ -1169,11 +1169,15 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
 	stmtCtx := logic.SCtx().GetSessionVars().StmtCtx
 	if stmtCtx.EnableOptimizeTrace {
 		tracer := &tracing.PhysicalOptimizeTracer{
-			PhysicalPlanCostDetails: make(map[int]*tracing.PhysicalPlanCostDetail),
+			PhysicalPlanCostDetails: make(map[string]*tracing.PhysicalPlanCostDetail),
 			Candidates:              make(map[int]*tracing.CandidatePlanTrace),
 		}
 		opt = opt.withEnableOptimizeTracer(tracer)
 		defer func() {
+			r := recover()
+			if r != nil {
+				panic(r) /* pass panic to upper function to handle */
+			}
```

Comment:
If we don't have panic here, why do we add this?

Reply:
we have run into a panic here. Steps as below:
1. Set the memory limit
2. Run trace plan, the memory usage exceed the limit, it will cause panic in the physical operator,  and run into line 1182, since some operation is breaked, so some paramater is null, it will run into null potiner panic, tidb get crashed.

Add recover here can let upper function handle out of memory panic to avoid crash.